src/pumipic_adjacency.tpp

```diff
@@ -544,5 +548,42 @@ namespace pumipic {
     Kokkos::Profiling::popRegion();
     return found;
   }
+
+  template <typename ParticleType, typename Segment3d>
+  struct NativeParticleHandlerAtElemBdry {
+    NativeParticleHandlerAtElemBdry () {}
+
+    void operator()(o::Mesh& mesh, ParticleStructure<ParticleType>* ptcls,
+                    o::Write<o::LO>& elem_ids, o::Write<o::LO>& inter_faces,
+                    o::Write<o::LO>& lastExit, o::Write<o::Real>& inter_points,
+                    o::Write<o::LO>& ptcl_done,
+                    bool requireIntersection,
+                    Segment3d x_ps_orig,
+                    Segment3d x_ps_tgt) const {
+      const auto side_is_exposed = mark_exposed_sides(&mesh);
+      // Check if intersection face is exposed
+      check_model_intersection(mesh, ptcls, x_ps_orig, x_ps_tgt, elem_ids,
+                               ptcl_done, lastExit, side_is_exposed,
+                               requireIntersection, inter_faces);
+
+      // Move to next element
+      set_new_element(mesh, ptcls, elem_ids, ptcl_done, lastExit);
```

Comment:
Is there ever a case where we would not call `set_new_element`? I think this should actually be outside of the functor and inside the `particle_search` function.

Reply:
I fully agree with this. I will remove it outside the functor.